text/0931-template-compiler-api.md

```diff
@@ -0,0 +1,431 @@
+      {
+        scope: (instance) => ({ 
+          Section, 
+          "#secret": instance.#secret
+        }),
+      },
+      this
+    )
+  }
+}
+```
+
+> This RFC is focused on making sure the scope accessors can do everything javascript can do, which is why we're including private fields. However, additional work beyond this RFC is required to make the template compiler correctly parse expressions like `{{this.#secret}}`. 
+
+#### Implicit Form
+
+The "Implicit Form" is cheaper and easier to produce because it doesn't need to do any lexical analysis and doesn't need to parse the handlebars at all.
+
+The downside is that data flow is not all statically visible, because it relies on `eval`.
+
+Implicit Form has two key use cases:
+ - as an intermediate format between a preprocessor stage (which can eliminate all GJS special syntax and semantics) and the rest of a standard Javascript toolchain.
+ - as the implementation format in sandbox-like environments where dynamic code execution is the whole point.
+
+Examples of Implicit Form:
+
+```js
+import { template } from '@ember/template-compiler';
+
+// Notice that all of these have the exact same
+// `params` argument. It's always the same. 
+// That's why it's easy to produce.
+
+const Headline = template(
+  "<h1>{{yield}}</h1>",
+  {
+    eval() { return eval(arguments[0]); }
+  }
+);
+
+const Section = template(
+  "<Headline>{{@title}}</Headline>", 
+  { 
+    eval() { return eval(arguments[0]); }
+  }
+);
+
+class extends Component {
+  static {
+    template(
+      "<Section @title={{this.title}} @subhead={{this.#secret}} />", 
+      {
+        // this handles private fields just fine,
+        // see Appendix A.
+        eval() { return eval(arguments[0]); }
+      },
+      this
+    )
+  }
+}
+```
+
+> _eval seems bad, what about Content Security Policy?_<br>
+> Typical apps never needs to actually _run_ the eval. This is a communication format between layers of build tooling. You _can_ run it, if you're making something like an interactive development sandbox. But that is a case that already requires `eval` anyway.
+
+> _Why `arguments[0]` instead of an explicit argument?_<br>
+> If we picked a local name to use for the argument, we would shadow that name in the surrounding scope. Whereas `arguments` is already a keyword that exists for this purpose, and can never collide with other local bindings.
+
+
+### Type Signature
+
+```ts
+import { ComponentLike } from '@glimmer/template';
+
+function template<S>(
+  templateContent: string,
+  params?: ExplicitParams | ImplicitParams,
+): ComponentLike<S>;
+function template<C extends ComponentLike<any>>(
+  templateContent: string,
+  params?: ExplicitParams | ImplicitParams,
+  backingClass: C;
+): C;
+
+interface BaseParams {
+  strict?: boolean;
+  moduleName?: string;
+}
+
+interface ExplicitParams extends BaseParams {
+  scope?: (instance: object) => Record<string, unknown>;
```

Comment:
Two notes on this type:

1. Note: the union will allow you to pass *both* `eval` and `scope`. [playground][1]. We have a choice to make here:

    - Do we indulge in [some type-level shenanigans](https://stackoverflow.com/a/52678379/564181) to forbid it?
    - Or do we just ship a lint rule or some such?

    I think either case is fine, since this is a low-level API which is mostly emitted by tools.

2. I think we want the callback signature to get an instance where there is a class, and no argument at all (or, for consistency with the current runtime behavior, `null`?) when there is no class passed?

Combining those two notes, I think our final signature here should look something like the one in [*this* playground][2], which makes the following changes:

1. It moves the class argument before the params to avoid the issue with optional params preceding required params.
2. It introduces a `StrictUnion` type to force excess property checking to do what we would expect so you *cannot* pass both `scope` and `eval`.
3. It updates the return type from `eval` to be `unknown` instead of `any`. While `eval()` itself returns `any`, this means our own code is stricter and safer here.
4. It passes no argument at all to the `scope` in the case of template-only components, and passes the instance type `scope` in the case of class-backed components.

[1]: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAcgAIHMA2wB2MD0BTTAN2CgkxAJgFo8QAjPKKtCCAZzwQG4AoH0SLDgBvOAGEI4MpQAywANZ44AXzgAzUiESoM2HDFpg0AQwPd+UoRKmZK6zdvSgKUHAGNJkW9nM8AJnhuJlBKagCumG4wwGRwBuAmBgA8AMoAfAAUPHBxhol4EtiUAFxwbDBQWCgANNlwYMZQxiBsAPylAKIAHkbAbsAwAAqNzWxwAD5wAJIJfQPDTS08AJSl1l6yCnipabwBQY2hEVExmLkJptticHhdBph+Y+vS2HKKScaYAJ5pmXXxRkuhXuMFK5UqmBqdQaizYnR6GH6QxGLQm01mSIWo1qOToxjc8iqYhMbDh4hWa14-CKUDU+KUACFjBwsajhHVwX0YO04HRWGg8J9eDkQBA-GEBQA5Zp4HmcyG8ZTUgy0+lwbq9TEosa3e6POBMlnakQcjxgWWlDJYcqfNx4UoQOgAK0CMGWcAAvGk4AAlQLQPxJeU1OAReSYCAAd0wux4SqwKrpdvRmvmxt1BH1hrwrLG7JyeEIxjQlvdXrgny+ir4APyGQABkkwGkABJ4NAsJI4Zv16omgtFtAZd35nJwEIwMJQM6F4sZRooMIUbBsADaAAYALrLYUqHFlM32uBWzA2yJ4Mve0c5CdTs7CJU5JXKHc8IA

[2]: https://www.typescriptlang.org/play?#code/PTAEFpM0GUEsDmA7AhgFwK4CcCmoAmOAZnEnGnAPZIRTiggBQjcAtgA6VZqgDkAAggA2pNMBxIAbnCzVWEtOBysARjizghlSgGccvANwsOXHgG9QAYUomkCgDJwA1ngC+oIrNZ9BIpGLRldiF0fSM2Tm4rG047fw8vH2E2eSxgAGMY6gVDZkJ0kNwPDCR0impQQI4QwIAeGAA+AApGUEqgmpxrfwUALlAdNCxSBAAaVtB2FCwUVh0Afn6YIbgygFUyalqAUQAPYNXyAAVp2Z0AMS4AFQ7QgHkkIQBPUAAfUABJasO0E5m5hqMACU-Wstgczhw9QaRnyhTwRBKZSoNCqwVCtUsoBwu0CSHwOmi4P8jhctRQSCeDWaEzRnW6eLQ-UGwyQYwmKhQ6ScI0sIR0OlB4zaU3+CyWK3WmyQO32InSx1Oc0uWD5KAFACEuS58JiGm9Pt8Fb8lTpASCrEYWD0sEQuXgtXo-mdQGYJizVmhFqAVNohDgKUY2qxKPgMP6AHKzHDej1soyuZiidR29J4PYHY3O5XXW6BB7PbG4iQE0COnDZwluto6TLsGP9JpA0AAXn1ACUcJksLq42NQCUnEhKAB3JAwxiJ62BW320AZ+WKsUqtWa7U4XVYnF40tg2IQskUqmAtrbkuE8uV13uusN0BNUiDClp-ofJBP0o4K5Pet65tt0BO27XsVjZUYByQIdR3HBMkxtVM8C+TMlxdM98QvdUK1Na9T0kFAhEbf99UHYcx1gxg0B-PANhRABpHAnh0Worn1FtQCuIsd0JDj5lAFwnkoIh2NAfo7EkdQjEo+tYElNAaOoAAJHAhHrLBmPAq4AEEhCEVjhLQ0sjwmXiOIAMlAP4KHw2ogK4XU9gKDBCFqeSkHoxjmO03TwP4wT2IacCSkIEg7HwakJlEnBxKwSSqJk4YpRRZi9OWBK5OlJSVPUdT-KtEBaCgeddlmYIcEJOh6CYRhMnfHgtHSfDWz4EcuCEfBckYRFSnKGgcQKFA2CaQUBlAhBm2rUBcEwLAaAAAwAdQAC0oFAAEJQAAEjMHRXFm8j8sAGXJGHy-g0B0JR9i7RR1FkLBaFAEc8AapBhx4KYBR9Sg0EWgZb1ACl8GxPChAAOjBii8xwJpZtqdgGkyrRwLMMx6vw1xXFW2pgDh2akYmKL8KbHC2kmnBpt64GmmmBAMHkfwdAAbQABgAXSBINQFcYVfsoetCNbfUmgsVGhE5oFxlcdnmHywBQcgh6pQmh2H4eUxHXRRygGqEdHMexhpceJ2teZwfmAKF0ARbFiWpeOsA5bpRWYbhhHKCRjWtZ1rGcbx3DCfGiY2im7AKcJ6naYURnWaltoucnG3Zeq-lCT2Er-QARk488iX3fxagsTSsAQYaLFQeRmVGgx-oQE3QCQWm1Du9xXH1Cbq54ABiQhaybcvWQQYnA7J4PQFmrbvrgHQQepyfS5wdwJ82sxx8n6ep+r9wngDLBCUoNqQb2iYpxrNB0FWAf2gVwIlbh5G+pCNhKkWieQc7sr0nRr39fA5efZJmtb0bI+E+n4iL3gmn-U8ux+psG5hAvgr9ay8H6EA58OAX5d3SLAkmkssGSw5omKcCd+ofRTtUHAAAmTO6Fs7ZFzvnQuxda7Rl7iMSuKBq6iXruoTmnMW4TDbqABB6Qe4jT7ufIOM0R5jyfivBhINZ7z0JNI5+q92Fz1AJvaYO894HxjswY+p90jnwdlfJ2DRb5QPvt4Ze6C34fz1gbH+59fZCCJuAiBEiQ6uLDnTM6zM2Ycz-rHIJ0dOaTmlmAA6909B4FengCos0CZCFmj6ZSo5E7qmTsVMhABmKhu4shxDQHnUABci79BLsw0RrCq41zrqobhTc+FtAEUIkRfZxFD0kaPJeMip5yIUaABeyjZFFzXuozR29QC73wPvfB+iRqGOMZDa+5izB3wGtYvpQj7He0fhPX+f8jZ83vCgkBAswEBwgRsmBVy-68CEUgoZtVUG2NrFgmO4s7kMDAKdc6OJ6xlCUFgW691HqgGeq9SYmTPrfR5tJAGQN8JgxBt8pJbjvmD3Jki7xhdw7038aEoJuDQkEMYEAA

Reply:
Thanks, this is great.

Reply:
I hit a snag with the idea to reorder the parameters. When the call arity is two, we need to distinguish the case where the second argument is a parameters bag vs a backing class. And I'd rather not make a lot of assumptions about what's inside a component. Maybe it even happens to have a property named scope, etc.

Perhaps instead we should move the backing class inside the params bag. Only downside I see is that it's slightly more wordy, since it will need a property name.

Reply:
That was the other thought I had. Otherwise, making the `params` argument required in both cases also works—and seems fine from the POV that this is a low-level API primarily for tools and secondarily for authors doing relatively low-level work.